FEAT support variables as second and third argument

lekoala opened the pull request:
## Description

This PR upgrades the text collector to support collecting arguments as variables for second and third argument (see examples below).

## Manual testing steps

```php
        // Should display last $str and support concatenation
        $str = 'Should NOT appear';
        $str = 'Lazy loading can increase perceived page performance by slightly delaying media loading. ' .
            'Eager loading will load files as soon as possible and can be used if the image is in view as the page ' .
            'loads (above or near the fold).';
        $titleTipContent = _t('DemoAdmin.LoadingTitleTip', $str);

        // Does not support variables, use {type}. This will improperly detect "test" as value for the translation
        $type = "test";
        $IDontHaveAValidSyntax = _t(__CLASS__ . '.CANNOT_DELETE', "Not allowed to delete '$type' records");

        // Support variable as third argument
        $var = ['some' => 'var'];
        $IWasNotCollected = _t('DemoAdmin.ImNOTCollected', 'test {some}', $var);

        // Support default case
        $ImCollected = _t('DemoAdmin.ImCollected', 'test {some}', ['some' => 'var']);
```

## Issues
- https://github.com/silverstripe/silverstripe-framework/issues/11268
- https://github.com/silverstripe/silverstripe-asset-admin/issues/1532

## Pull request checklist
- [x] The target branch is correct
    - See [picking the right version](https://docs.silverstripe.org/en/contributing/code/#picking-the-right-version)
- [x] All commits are relevant to the purpose of the PR (e.g. no debug statements, unrelated refactoring, or arbitrary linting)
    - Small amounts of additional linting are usually okay, but if it makes it hard to concentrate on the relevant changes, ask for the unrelated changes to be reverted, and submitted as a separate PR.
- [x] The commit messages follow our [commit message guidelines](https://docs.silverstripe.org/en/contributing/code/#commit-messages)
- [x] The PR follows our [contribution guidelines](https://docs.silverstripe.org/en/contributing/code/)
- [x] Code changes follow our [coding conventions](https://docs.silverstripe.org/en/contributing/coding_conventions/)
- [x] This change is covered with tests (or tests aren't necessary for this change)
- [ ] Any relevant User Help/Developer documentation is updated; for impactful changes, information is added to the changelog for the intended release
- [x] CI is green

lekoala committed:
FEAT support variables as second and third argument

lekoala committed:
revert invalid formatting

GuySartorelli commented:
Please fill in the PR checklist. That checklist is there to help you as the contributor double check you've done everything you need to do. That means that I as the reviewer will then know whether you have done those things which speeds things up a little.

GuySartorelli commented on src/i18n/TextCollection/i18nTextCollector.php:
```diff
@@ -618,15 +618,31 @@ public function collectFromCode($content, $fileName, Module $module)
+                if ($id === T_VARIABLE) {
+                    $inVar = $text;
+                }
+                if ($id === T_CONSTANT_ENCAPSED_STRING && $inVar && $text) {
```
Are falsy checks correct here? What if the values are an empty string or `'0'`?

lekoala replied:
mhh an empty string doesn't matter, but indeed "0" as a translation would fail at the moment
not very likely that someone would do $str = '0' but still i guess I could improve this :)

lekoala replied:
updated

GuySartorelli commented on src/i18n/TextCollection/i18nTextCollector.php:
```diff
@@ -618,15 +618,31 @@ public function collectFromCode($content, $fileName, Module $module)
         $currentUse = null;
         $currentUseAlias = null;
+
+        $inVar = null;
```
What is "inVar"? Does this just track "something is in a variable"?
It looks like we use it to track only the most recently encountered variable in the entire file, but only until the assignment is finished being parsed and then we stop tracking it. Is that correct?
Might warrant an inline comment.

lekoala replied:
inVar is how i've collected $str = 'something' before the _t call is made
this allows         $str = 'Lazy loading can increase perceived page performance by slightly delaying media loading. ' .
            'Eager loading will load files as soon as possible and can be used if the image is in view as the page ' .
            'loads (above or near the fold).';
        $titleTipContent = _t('DemoAdmin.LoadingTitleTip', $str);
to work, even when being multiple strings concatenated together

GuySartorelli replied:
Does it correctly work for adding to a variable later? e.g.
```php
$x = 'some value';
$x .= ' and some more text';
```

GuySartorelli commented on src/i18n/TextCollection/i18nTextCollector.php:
```diff
@@ -618,15 +618,31 @@ public function collectFromCode($content, $fileName, Module $module)
+                    $stringVariables[$inVar] = $text;
+                    $inVar = null;
+                }
+                if ($text === ';') {
```
```suggestion
                // Stop tracking the variable as we've finished parsing its assignment
                if ($text === ';') {
```
This assumes my earlier assumption about what `$inVar` is used for is correct.

lekoala replied:
yes

GuySartorelli commented on src/i18n/TextCollection/i18nTextCollector.php:
```diff
@@ -618,15 +618,31 @@ public function collectFromCode($content, $fileName, Module $module)
         $potentialClassName = null;
         $currentUse = null;
         $currentUseAlias = null;
+
```
```suggestion
```
Doesn't seem like there's any reason to separate these variables from the ones above.

lekoala replied:
it's just a way to "group" them visually but i'll happily remove the extra space

GuySartorelli commented on src/i18n/TextCollection/i18nTextCollector.php:
```diff
@@ -618,15 +618,31 @@ public function collectFromCode($content, $fileName, Module $module)
         $currentUseAlias = null;
+
+        $inVar = null;
+        $stringVariables = [];
```
Are we tracking _all_ variables that contain a string in the whole file? That's probably okay given the way this parsing is done, just want to be sure I understand what's going on.

lekoala replied:
yes, i track everything because you don't know what is going to be called later on

GuySartorelli commented on src/i18n/TextCollection/i18nTextCollector.php:
```diff
@@ -733,6 +749,28 @@ public function collectFromCode($content, $fileName, Module $module)
                     continue;
                 }
 
+                // Allow _t(Entity.Key, 'translation', $var) and expand _t(Entity.Key, $var)
```
What part of these examples is the part being allowed? And what do you mean by "expand" the second example?

GuySartorelli replied:
Please respond to these questions.

GuySartorelli replied:
Nevermind

GuySartorelli commented on src/i18n/TextCollection/i18nTextCollector.php:
```diff
@@ -733,6 +749,28 @@ public function collectFromCode($content, $fileName, Module $module)
 
+                // Allow _t(Entity.Key, 'translation', $var) and expand _t(Entity.Key, $var)
+                if ($id == T_VARIABLE && !empty($currentEntity)) {
+                    // We have a translation, eg: _t(Entity.Key, 'translation', $var)
+                    if (count($currentEntity) == 2) {
+                        continue;
+                    }
```
How does having exactly two items mean we have a translation? Are the two items the key and the default text?
And is the continue because it doesn't matter what's in any subsequent arguments to the `_t()` function?

lekoala replied:
the current entity is an array with the following entries: name, translation, comments

if count = 1 : the translation key is being defined, but no translation has been found yet
if count = 2 : a translation has been stored for this key
if count = 3 : there is a comment

to be honest it's not very intuitive and it would deserve to be refactored to something less arbitrary (some kind of structered data) but that's how it works at the moment and i didn't want to change everything

GuySartorelli replied:
Okay. My main confusion was with the word "translation" - this isn't a translation, it's just some default text with a key. Translations come separately.

GuySartorelli commented on src/i18n/TextCollection/i18nTextCollector.php:
```diff
@@ -733,6 +749,28 @@ public function collectFromCode($content, $fileName, Module $module)
+                        continue;
+                    }
+
+                    // The variable is the second argument or present in the parameter
```
What is "the parameter" in this context? Does that mean it's in the third (or subsequent) argument?

lekoala replied:
it's when you have _t('key', 'translation', $var) or _t('key', 'translation', 'comments' . $var) => not than this should be ever done anyway, but that can happen

GuySartorelli replied:
So you mean it's the third argument (or after the second argument - I guess maybe you mean it could be in the third or later argument?)? If so please change the comment to specify that.

lekoala replied:
i'm not sure how to phrase that correctly :-)

GuySartorelli replied:
Reading again I think it's probably fine as is

GuySartorelli commented on src/i18n/TextCollection/i18nTextCollector.php:
```diff
@@ -733,6 +749,28 @@ public function collectFromCode($content, $fileName, Module $module)
 
+                // Allow _t(Entity.Key, 'translation', $var) and expand _t(Entity.Key, $var)
+                if ($id == T_VARIABLE && !empty($currentEntity)) {
+                    // We have a translation, eg: _t(Entity.Key, 'translation', $var)
```
By "we have a translation" do you mean "we've fully captured this localisation key and its default text"? If so, please change the comment to something in that vain so it's clear what you mean.

lekoala replied:
We have a translation just means the second argument has been provided and that we can proceed

GuySartorelli replied:
The second argument isn't a translation, it's just the default text to display. Please update the comment to say something like "We have found the default text for this key"

lekoala replied:
done

GuySartorelli commented on src/i18n/TextCollection/i18nTextCollector.php:
```diff
@@ -733,6 +749,28 @@ public function collectFromCode($content, $fileName, Module $module)
+                    // Try to find it _t(Entity.Key, $var)
+                    $stringValue = $stringVariables[$text] ?? null;
+
+                    // It's translated, continue
```
By "It's translated" do you mean "We've fully captured this localisation key and its default text"? If so, please change the comment to something in that vain so it's clear what you mean.

lekoala replied:
it's translated in this case means a translation has been provided and need to be processed. by default captured string look like 'my translation' and need to go through processString before being stored

GuySartorelli replied:
> it's translated in this case means a translation has been provided

How does text collector know there are translations provided? It shouldn't be aware of the `lang/*.yml` files.

Edit: Sounds like you've been using "translation" to refer to the default text (i.e. second argument)?

lekoala replied:
yes indeed, i've updated the comments

GuySartorelli commented on src/i18n/TextCollection/i18nTextCollector.php:
```diff
@@ -733,6 +749,28 @@ public function collectFromCode($content, $fileName, Module $module)
+                    // It's translated, continue
+                    if ($stringValue) {
+                        $stringValue = $this->processString($stringValue);
+                        // Deal with ''
```
Does this mean "deal with empty strings" or "deal with single quotes"? OR something else?
And in either case, what scenario do we find those in? Is it in the first argument? The second? Somewhere else?

lekoala replied:
it means if a string was an empty string (=> '') then there is no point in storing an empty translation

GuySartorelli reviewed (changes requested):
Most of this review is just me asking for clarification, since I haven't worked in this part of the code much before so I need a bit of help understanding what's going on.

GuySartorelli commented:
Please retarget this to the `5` branch - it's adding new functionality and has a risk of regression that I'd rather not introduce in a patch.

lekoala committed:
Update src/i18n/TextCollection/i18nTextCollector.php

Co-authored-by: Guy Sartorelli <36352093+GuySartorelli@users.noreply.github.com>

lekoala commented:
@GuySartorelli ideally i'd like to make sure there is no regression
currently, code like this  https://github.com/silverstripe/silverstripe-asset-admin/issues/1532 is not being collected so I assume it is considered "valid code" according to current features, no?

GuySartorelli commented:
Any change has some chance of regression. There's no _known regressions_ this causes, but because it's code I'm unfamiliar with, there is a chance of regressions that I won't be able to identify.

Regardless, as this is adding new functionality, it is better included in a minor release rather than a patch release.

lekoala commented:
ok rebased!

i don't know if any test or update is needed

a unit test would be nice but i didn't go as far as that. i've just tried locally on my project and it seemed to work fine

GuySartorelli commented:
I won't have time to test this out myself until at least halfway through next week.
Unit tests would be great if you have time to implement them.

lekoala commented:
@GuySartorelli sure (that made me realize i was missing a line in my code^^)

This is on a blank install on my machine

Without cache folder already created (initial warm up, or if clearing the whole cache directory)

manifest
---
Time: 8s
Peak memory: 16.000Mb
Memory: 16.00Mb

improved_manifest
---
Time: 1s
Peak memory: 16.000Mb
Memory: 16.00Mb

With the case already in place

manifest
---
Time: 619ms
Peak memory: 6.000Mb
Memory: 6.00Mb

improved_manifest
---
Time: 417ms
Peak memory: 6.000Mb
Memory: 6.00Mb


With no surprise, creating hundreds of files on the hard drive takes a substantial amount of time on the first run. Having a in memory array also helps a lot. This comes at almost no performance cost since the whole "manifest" array end up loaded anyway, and it's larger than the file cache.

This could certainly be improved even more (i don't see the need of having a separate file cache, the whole thing could fit in the same array) and the array could be optimized to avoid storing empty entries (which is, let's face it, most of it!)

lekoala committed:
add test for variables as second and third arguments

lekoala committed:
refactor string variables handling

lekoala committed:
move comment to docblock

lekoala commented:
@GuySartorelli i added the tests and improved how i track strings. i added some comments to make it more clear how it works. the whole test suite is running fine locally

lekoala committed:
no need to call processString again

GuySartorelli commented on src/i18n/TextCollection/i18nTextCollector.php:
```diff
@@ -618,15 +618,55 @@ public function collectFromCode($content, $fileName, Module $module)
         $potentialClassName = null;
         $currentUse = null;
         $currentUseAlias = null;
+        $inVar = null;
```
```suggestion
        $inVar = null; // Tracks the value currently being assigned to a variable
```
As per https://github.com/silverstripe/silverstripe-framework/pull/11521#discussion_r1903932492
Have I understood that correctly? If not please adjust the comment so it's appropriately indicating the purpose of this variable.

lekoala replied:
it's not tracking the value, it's tracking the name of the variable. the value is stored in inVarText. i've added comments to reflect this

GuySartorelli commented on src/i18n/TextCollection/i18nTextCollector.php:
```diff
@@ -618,15 +618,55 @@ public function collectFromCode($content, $fileName, Module $module)
+                        $inVarText = '';
+                        continue;
+                    }
+                    if ($id === T_CONSTANT_ENCAPSED_STRING && $inVar && $text) {
```
```suggestion
                    if ($id === T_CONSTANT_ENCAPSED_STRING && $inVar !== null && $text !== null) {
```
As per https://github.com/silverstripe/silverstripe-framework/pull/11521#discussion_r1903933994

lekoala replied:
perfect

lekoala committed:
update comment

lekoala committed:
update comment

lekoala committed:
explain inVar and inVarText

lekoala committed:
Update src/i18n/TextCollection/i18nTextCollector.php

Co-authored-by: Guy Sartorelli <36352093+GuySartorelli@users.noreply.github.com>

lekoala committed:
add new test cases

lekoala committed:
support falsy values and .= operator

lekoala commented:
@GuySartorelli updated the code and added test cases for falsy values and $x = 'test' ; $x .= 'more';

GuySartorelli reviewed (approved):
Looks good, great unit test. Thank you!
